tftp/tftp.1.in
.\" -*- nroff -*- --------------------------------------------------------- *
.\"  
.\" Copyright (c) 1990, 1993, 1994
.\"     The Regents of the University of California.  All rights reserved.
.\"
.\" Copyright 2001 H. Peter Anvin - All Rights Reserved
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"----------------------------------------------------------------------- */
.TH TFTP 1 "23 July 2008" "tftp-hpa @@VERSION@@" "User's Manual"
.SH NAME
.B tftp
\- IPv4 Trivial File Transfer Protocol client
.SH SYNOPSIS
.B tftp
[ \fIoptions...\fP ]
[\fIhost\fP [\fIport\fP]]
[\fB\-c\fP \fIcommand\fP]
.br
.SH DESCRIPTION
.B tftp
is a client for the Trivial file Transfer Protocol, which can be
used to transfer files to and from remote machines, including some
very minimalistic, usually embedded, systems.  The remote
.I host
may be specified on the command line, in which case
.B tftp
uses
.I host
as the default host for future transfers (see the
.B connect
command below.)
.SH OPTIONS
.TP
.B \-4
Connect with IPv4 only, even if IPv6 support was compiled in.
.TP
.B \-6
Connect with IPv6 only, if compiled in.
.TP
\fB\-c\fP \fIcommand\fP
Execute \fIcommand\fP as if it had been entered on the tftp prompt.
Must be specified last on the command line.
.TP
.B \-l
Default to literal mode. Used to avoid special processing of ':' in a
file name.
.TP
\fB\-m\fP \fImode\fP
Set the default transfer mode to \fImode\fP.  This is usually used with \-c.
.TP
\fB\-R\fP \fIport:port\fP
Force the originating port number to be in the specified range of port
numbers.
.TP
.B \-v
Default to verbose mode.
.TP
.B \-V
Print the version number and configuration to standard output, then
exit gracefully.
.TP
.B \-B\fP \fIblocksize\fP
Set the "blocksize" TFTP option (RFC 2348) to the specified value.
.TP
.B \-w\fP \fIwindowsize\fP
Set the "windowsize" TFTP option (RFC 7440) to the specified value.
.SH COMMANDS
Once
.B tftp
is running, it issues the prompt
\f(CWtftp>\fP
and recognizes the following commands:
.TP
\fB?\fP \fIcommand-name...\fP
.TP
\fBhelp\fP \fIcommand-name...\fP
Print help information
.TP
.B ascii
Shorthand for
.BR "mode ascii" .
.TP
.B binary
Shorthand for
.BR "mode binary" .
.TP
\fBconnect\fP \fIhost [port]\fP
Set the
.I host
(and optionally
.IR port )
for transfers.  Note that the TFTP protocol, unlike the FTP protocol,
does not maintain connections between transfers; thus, the
.B connect
command does not actually create a connection, but merely remembers
what host is to be used for transfers.  You do not have to use the
.B connect
command; the remote host can be specified as part of the
.B get
or
.B put
commands.
.TP
\fBget\fP \fIfile\fP
.sp -.6l
.TP
\fBget\fP \fIremotefile localfile\fP
.sp -.6l
.TP
\fBget\fP \fIfile1 file2 file3...\fP
Get a file or set of files from the specified sources.  A remote
filename can be in one of two forms: a plain filename on the remote
host, if the host has already been specified, or a string of the form
.I "host:filename"
to specify both a host and filename at the same time.  If the latter
form is used, the last hostname specified becomes the default for
future transfers.  Enable
.B literal
mode to prevent special treatment of the ':' character (e.g. C:\\dir\\file).
.TP
.B literal
Toggle literal mode.  When set, this mode prevents special treatment of ':' in filenames. 
.TP
\fBmode\fP \fItransfer-mode\fP
Specify the mode for transfers;
.I transfer-mode
may be one of
.B ascii
(or
.BR netascii )
or
.B binary
(or
.BR octet .)
The default is
.BR ascii .
.TP
\fBput\fP \fIfile\fP
.sp -.6l
.TP
\fBput\fP \fIlocalfile remotefile\fP
.sp -.6l
.TP
\fBput\fP \fIfile1 file2 file3... remote-directory\fP
Put a file or set of files to the specified remote file or directory.
The destination can be in one of two forms: a filename on the remote
host, if the host has already been specified, or a string of the form
.I "host:filename"
to specify both a host and filename at the same time.  If the latter
form is used, the hostname specified becomes the default for future
transfers.  If the remote-directory form is used, the remote host is
assumed to be a UNIX system or another system using
.B /
as directory separator.  Enable
.B literal
mode to prevent special treatment of the ':' character (e.g. C:\\dir\\file).
.TP
.B quit
Exit
.BR tftp .
End-of-file will also exit.
.TP
\fBrexmt\fP \fIretransmission-timeout\fP
Set the per-packet retransmission timeout, in seconds.
.TP
.B status
Show current status.
.TP
\fBtimeout\fP \fItotal-transmission-timeout\fP
Set the total transmission timeout, in seconds.
.TP
.B trace
Toggle packet tracing (a debugging feature.)
.TP
.B verbose
Toggle verbose mode.
.SH "NOTES"
The TFTP protocol provides no provisions for authentication or
security.  Therefore, the remote server will probably implement some
kinds of access restriction or firewalling.  These access restrictions
are likely to be site- and server-specific.
.SH "AUTHOR"
This version of
.B tftp
is maintained by H. Peter Anvin <hpa@zytor.com>.  It was derived from,
but has substantially diverged from, an OpenBSD source base, with
added patches by Markus Gutschke and Gero Kulhman.
.SH "SEE ALSO"
.BR tftpd (8).


changelog.txt
commit ed30f904ea00e36a535fe8511d27f3603171bca6
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Sun Dec 8 13:50:52 2019 +0100

    fix common die() function usage

    update TODO! max blocksize seems to be 9212 bytes

commit 0c7400c4626ed8be9c73053533f10ff1c9880b8e
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Sun Dec 8 10:06:50 2019 +0100

    do not abort when unsuported options are requested

    NOTE: in netascii mode, tsize option can't be calculated

commit 86a0d4169390fd7dc88c8507dd8776be65787b4a
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Sun Dec 8 08:56:01 2019 +0100

    generate header dependency files and use them

    Note: the include order should not matter!
    reoder includes, force header to be compileable

commit cacdce80c1b97e33d1f0743e20ee2498be3e9da1
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Sat Dec 7 20:01:27 2019 +0100

    use tsize option too

    show used options with status command too

commit 23fc685e956c8e4da5df22e3736a96f8195a30c8
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Sat Dec 7 11:02:58 2019 +0100

    fix option handling

    write newline too with printf() messages

commit 9e9051555615860f68544a96fef13aa94b529563
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Sat Dec 7 10:07:02 2019 +0100

    add blocksize option to tftp client too

    replace more tabs and tailing white spaces

commit 1784d176830f1b275f5949869e2f5ee5bb41dcc6
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Thu Dec 5 20:54:35 2019 +0100

    expand tabs

commit 1d165aaf66fb084d70ff419d3091bbdcc19eef73
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Mon Nov 25 23:05:56 2019 +0100

    prevent more warnings on mac OSX

commit da3334ae24e60b13598a946959d1e7670abfd42c
Author: Claus Klein <claus.klein@arcormail.de>
Date:   Mon Nov 25 14:23:16 2019 +0100

    be more strict, prevent gcc -Wextra warnings

    prevent buffer overflow too

commit c672c64fcd4c00d5c47af74ac67bcca07c9d8c8d
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Mon Nov 25 10:35:41 2019 +0100

    added new files from ReadHead patches too

commit f4958d94141bcff88c432f4eda838603a1c20597
Author: ClausKlein <claus.klein@arcormail.de>
Date:   Mon Nov 25 10:30:16 2019 +0100

    Support for the TFTP windowsize option (RFC 7440)

    see https://git.centos.org/forks/clausklein/rpms/tftp.git

    apply patches from Red Hat 7.5:
    tftp-0.40-remap.patch
    tftp-hpa-0.39-tzfix.patch
    tftp-0.42-tftpboot.patch
    tftp-0.49-chk_retcodes.patch
    tftp-hpa-0.49-fortify-strcpy-crash.patch
    tftp-0.49-cmd_arg.patch
    tftp-hpa-0.49-stats.patch
    tftp-hpa-5.2-pktinfo.patch
    tftp-doc.patch
    tftp-enhanced-logging.patch
    tftp-rfc7440-windowsize.patch
    tftp-rewrite-macro.patch


tftp/tftp.c
/*
 * Copyright (c) 1983, 1993
 *      The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "extern.h"

#include <stdarg.h>
#include <poll.h>

static char pktbuf[PKTSIZE];

static void printstats(const char *, unsigned long);
static void startclock(void);
static void stopclock(void);
static void timed_out(void)
{
    printf("client: timed out\n");
    exit(1);
}

static size_t make_request(unsigned short opcode,
                           const char *name,
                           const char *mode,
                           size_t blocksize,
                           int windowsize,
                           size_t tsize,
                           struct tftphdr *out)
{
    char *cp, buf[16];
    size_t len;

    cp = (char *)&(out->th_stuff);

    out->th_opcode = htons(opcode);

    len = strlen(name) + 1;
    memcpy(cp, name, len);
    cp += len;

    len = strlen(mode) + 1;
    memcpy(cp, mode, len);
    cp += len;

    /* Don't include options with default values. */

    if (blocksize != SEGSIZE) {
        len = strlen("blksize") + 1;
        memcpy(cp, "blksize", len);
        cp += len;
        if (snprintf(buf, 16, "%lu", blocksize) < 0)
            die("out of memory");
        len = strlen(buf) + 1;
        memcpy(cp, buf, len);
        cp += len;
    }

    if (windowsize > 0) {
        len = strlen("windowsize") + 1;
        memcpy(cp, "windowsize", len);
        cp += len;
        if (snprintf(buf, 16, "%u", windowsize) < 0)
            die("out of memory");
        len = strlen(buf) + 1;
        memcpy(cp, buf, len);
        cp += len;
    }

    {
        len = strlen("tsize") + 1;
        memcpy(cp, "tsize", len);
        cp += len;
        if (snprintf(buf, 16, "%zu", tsize) < 0)
            die("out of memory");
        printf("client: option request tsize:%s\n", buf);
        len = strlen(buf) + 1;
        memcpy(cp, buf, len);
        cp += len;
    }

    return (cp - (char *)out);
}

static void send_request(int sock,
                         union sock_addr *to,
                         short request,
                         const char *name,
                         const char *mode,
                         size_t blocksize,
                         unsigned windowsize,
                         size_t tsize)
{
    struct tftphdr *out;
    size_t size;

    out = (struct tftphdr *)pktbuf;
    size = make_request(request, name, mode, blocksize, windowsize, tsize, out);

    if (sendto(sock, out, size, 0, &to->sa, SOCKLEN(to)) != (unsigned)size)
        die("send_request: sendto: %s", strerror(errno));
}

static int wait_for_oack(int sock, union sock_addr *from, char **options, int *optlen)
{
    unsigned short in_opcode;
    struct tftphdr *in;
    int r;

    r = recvfrom_with_timeout(sock, pktbuf, sizeof(pktbuf), from, TIMEOUT);
    if (r == 0)
        return r;

    in = (struct tftphdr *)pktbuf;
    in_opcode = ntohs(in->th_opcode);

    if (in_opcode == ERROR)
        die_on_error(in);
    if (in_opcode != OACK)
        return -in_opcode;

    *options = pktbuf + 2;
    *optlen = r - 2;

    return 1;
}

static size_t get_tsize(int fd) {
    struct stat stbuf = {};
    size_t tsize = 0;
    if (fstat(fd, &stbuf) >= 0) {
        tsize = stbuf.st_size;
    } else {
        perror("fstat()");
    }

    return tsize;
}

struct option_values {
    int windowsize;
    size_t blocksize;
    size_t tsize;
};

static int tftp_parse_oack(int sock, union sock_addr *from, struct option_values *def)
{
    char *options;
    int optlen;
    int retries = RETRIES;
    int r;
    do {
        r = wait_for_oack(sock, from, &options, &optlen);
        if (r < 0) {
            return 0;   // ACK received! CK
        }

        /* Parse returned options. */
        int n = 0;
        if (r != 0) {
            char *opt, *val;

            while (n < optlen) {
                opt = options + n;
                n += strlen(opt) + 1;
                val = options + n;

                if (str_equal(opt, "windowsize") && def->windowsize != 1) {
                    int v = atoi(val);  // TODO Use strtoul(() or strtoumax()
                    if (v != def->windowsize)
                        printf("client: server negotiated different windowsize: %d\n", v);
                    /* FIXME Assumes v > 0, it probably shouldn't. */
                    def->windowsize = v;
                }

                if (str_equal(opt, "blocksize") && def->blocksize != SEGSIZE) {
                    size_t bs = strtoul(val, NULL, 10);  // TBD Use strtoumax()
                    if (bs != def->blocksize)
                        printf("client: server negotiated different blocksize: %zu\n", bs);
                    if ((bs < SEGSIZE) || (bs > MAX_SEGSIZE))
                        die("client: server didn't negotiate valid blocksize");

                    def->blocksize = bs;
                }

                if (str_equal(opt, "tsize")) {
                    size_t ts = strtoumax(val, NULL, 10);
                    printf("client: server negotiated tsize: %zu\n", ts);
                    if (ts != def->tsize)
                        def->tsize = ts;
                }

                n += strlen(val) + 1;
            }

            if (def->windowsize < 1) {
                def->windowsize = 1;
                printf("client: server didn't negotiate windowsize, continuing with windowsize=1\n");
            }
        }
    } while (r == 0 && --retries > 0);

    if (retries <= 0)
        timed_out();

    return 1;   // OK
}


/*
 * Send the requested file.
 */
void tftp_sendfile(int fd, const char *name, const char *mode, int windowsize)
{
    union sock_addr server = g_peeraddr;
    unsigned long amount = 0;
    size_t blocksize = g_blocksize;
    FILE *fp;
    size_t tsize = get_tsize(fd);  // TODO only useable for mode == "octet" ! CK

    set_verbose(g_trace_opt + g_verbose);

    startclock();
    send_request(g_s, &server, WRQ, name, mode, blocksize, windowsize, tsize);

#if 0
    /* If no windowsize was specified on the command line,
     * don't bother with options.
     * When blocksize is supported, this should actually only be called
     * if no options were sent in the WRQ.
     */
    if ((windowsize < 0) && (blocksize == SEGSIZE))
        goto no_options;
#endif

    struct option_values values = {windowsize, blocksize, tsize};
    int ok = tftp_parse_oack(g_s, &server, &values);
    if (!ok) return;

#if 0
//XXX no_options:
    if (!ok) {
        int r;
        struct tftphdr *tp = (struct tftphdr *)pktbuf;
        int retries = RETRIES;
        do {
            r = recvfrom_with_timeout(g_s, pktbuf, sizeof(pktbuf), &server, TIMEOUT);
            if (r == 0) {
                /* Timed out. */
                continue;
            }
            if (ntohs(tp->th_opcode) == ACK && ntohs(tp->th_block) == 0) {
                break;
            } else if (ntohs(tp->th_opcode) == ERROR) {
                die_on_error(tp);
            }
        } while (r == 0 && --retries > 0);
        if (retries <= 0)
            timed_out();
    }
#endif

    fp = fdopen(fd, "r");
    int r = sender(g_s, &server, values.blocksize, values.windowsize, TIMEOUT, 0, fp, &amount);
    if (r < 0)
        exit(1);

    stopclock();
    if (amount > 0)
        printstats("Sent", amount);
    fclose(fp);
}

/*
 * Receive a file.
 */
void tftp_recvfile(int fd, const char *name, const char *mode, int windowsize)
{
    union sock_addr server = g_peeraddr;
    unsigned long amount = 0;
    size_t blocksize = g_blocksize;
    char error[ERROR_MAXLEN] = {};
    FILE *fp;
    size_t tsize = 0;

    set_verbose(g_trace_opt + g_verbose);

    startclock();
    send_request(g_s, &server, RRQ, name, mode, blocksize, windowsize, tsize);

#if 0
    /* If no windowsize was specified on the command line,
     * don't bother with options.
     * When blocksize is supported, this should actually only be called
     * if no options were sent in the RRQ.
     */
    if ((windowsize < 0) && (blocksize == SEGSIZE))
        goto no_options;
#endif

    struct option_values values = {windowsize, blocksize, tsize};
    int ok = tftp_parse_oack(g_s, &server, &values);
    if (!ok) return;

#if 0
//XXX no_options:
    if (!ok) {
        int r;
        struct tftphdr *tp = (struct tftphdr *)pktbuf;
        int retries = RETRIES;
        do {
            r = recvfrom_with_timeout(g_s, pktbuf, sizeof(pktbuf), &server, TIMEOUT);
            if (r == 0) {
                /* Timed out. */
                continue;
            }
            if (ntohs(tp->th_opcode) == ACK && ntohs(tp->th_block) == 0) {
                break;
            } else if (ntohs(tp->th_opcode) == ERROR) {
                die_on_error(tp);
            }
        } while (r == 0 && --retries > 0);
        if (retries <= 0)
            timed_out();
    }
#endif

    send_ack(g_s, &server, 0);

    fp = fdopen(fd, "w");
    int r = receiver(g_s, &server, values.blocksize, values.windowsize, TIMEOUT, fp, &amount, error);
    if (r < 0) {
        fprintf(stderr, "client: %s\n", error);
        exit(1);
    }

    stopclock();
    if (amount > 0)
        printstats("Received", amount);
    fclose(fp);
}

struct timeval tstart;
struct timeval tstop;

static void startclock(void)
{
    (void)gettimeofday(&tstart, NULL);
}

static void stopclock(void)
{

    (void)gettimeofday(&tstop, NULL);
}

static void printstats(const char *direction, unsigned long amount)
{
    double delta;

    delta = (tstop.tv_sec + (tstop.tv_usec / 1000000.0)) -
        (tstart.tv_sec + (tstart.tv_usec / 1000000.0));
    if (g_verbose) {
        printf("%s %lu bytes in %.1f seconds\n", direction, amount, delta);
        /* TODO: Change the statistics in a separate patch (bits???)! */
        printf(" [%.0f bit/s]\n", (amount * 8.) / delta);
        putchar('\n');
    }
}

